Add go.mod file

SamWhited opened the pull request:
Hello, please consider adding this go.mod file and then creating a tagged release. Just like in #4 I am having this be pulled in as an indirect dependency and would like to be able to keep track of a specific version instead of using an indirect import like:

> github.com/shurcooL/sanitized_anchor_name v0.0.0-20170918181015-86672fcb3f95 // indirect

To be clear: this isn't causing it any problems, I'm just trying to make my `go.mod` file easier to read by getting more projects that I indirectly depend on to support Go Modules.

Thank you for your consideration.

SamWhited committed:
Add go.mod file

SamWhited commented on go.mod:
```diff
@@ -0,0 +1,3 @@
+module github.com/shurcooL/sanitized_anchor_name
+
+go 1.9
```
I don't know what version you officially support, but I tested with 1.9.7 and everything appeared to work. 1.10 is the last supported version of go, so maybe you'd rather I bump this to that?

dmitshur replied:
Where is this "go" directive documented? What are the consequences of using it? I'd rather avoid adding it until I understand the implications.

SamWhited replied:
As far as I know it's not documented anywhere, but running `go mod init` creates a mod file with this line using whatever version of Go you're currently using.

dmitshur replied:
I see, thanks. Let's leave it out for now since it's not mandatory, and consider adding it when we have a better understanding of it.

SamWhited replied:
Since it's put in by the Go tool by default, shouldn't we leave it in and not remove it unless we understand why? It makes me nervous to do things that aren't the default thing recommended by the tooling.

This will only affect behavior (if it does so at all) on versions of Go using Go Modules, so it's not as if it could break older versions that don't support it.

dmitshur replied:
What version of Go did you use? When I do `go mod init` with an explicit module path, I don't get a `go` directive in it. See:

```
$ go version
go version go1.11.2 darwin/amd64
$ go mod init github.com/shurcooL/sanitized_anchor_name
go: creating new go.mod: module github.com/shurcooL/sanitized_anchor_name
$ ls
go.mod
$ cat go.mod 
module github.com/shurcooL/sanitized_anchor_name
```

Also, examples of go.mod files without a `go` directive are shown at https://golang.org/cmd/go/#hdr-The_go_mod_file, https://research.swtch.com/vgo-module, and https://github.com/golang/go/wiki/Modules.

SamWhited replied:
Oops, replied in the wrong place apparently. I ended up finding documentation for this here: https://tip.golang.org/cmd/go/#hdr-The_go_mod_file

So presumably it's still only in 1.12

dmitshur replied:
Thanks. There's not a lot of information there to give me an understanding of the implications of using that directive and its different values.

There's more information about it at https://github.com/golang/go/issues/28221#issuecomment-432765035.

I prefer not to add a `go` directive to `go.mod` for now. I will consider adding it later when it's clear what the benefits and consequences are.

Go 1.11.4 is the current stable Go release, and running `go mod init` with it produces a `go.mod` file with only the `module` directive. Go 1.12 is not yet released, it's in beta, so in theory anything there can change before the final release. I don't expect it will, my point is that it's clearly valid to rely on a `go.mod` generated by 1.11.4.

So, please remove the `go` directive and I will be able to merge this PR. Thanks.

dmitshur commented:
Thank you for the PR and for elaborating on the motivation for wanting this, and how it affects you.

It's reasonable to mark the current `sanitized_anchor_name` as v1.0.0. It's very stable and hasn't changed in a long time.

I'll make this change.

SamWhited commented:
It only adds it by default in recent commits; I don't remember exactly when it was added but on recent builds that will be in 1.12 it should do it.

dmitshur reviewed (changes requested):
Please make the change described in https://github.com/shurcooL/sanitized_anchor_name/pull/6#discussion_r243751450 and I'll be able to merge this. Thank you.

SamWhited commented:
Done

dmitshur commented:
The change I requested was to remove the go directive, but the latest commit changes its value from 1.9 to 1.12 instead. I'll make the change myself to save on the extra latency introduced by round-tripping via review comments. Thank you.

dmitshur committed:
Remove go directive from go.mod.

It's not something that a stable version of Go generates, 
so avoid adding it for now.

SamWhited commented:
Oops, sorry, this was not intentional. Like I said, the tooling keeps adding it back (in a recent build from master anyways).

dmitshur commented:
I understand. Thank you again!

dmitshur commented:
@SamWhited I've made a [v1.0.0](https://github.com/shurcooL/sanitized_anchor_name/releases/tag/v1.0.0) tag, so anyone using this package with Go modules should see something like this in their go.mod file now:

```
require github.com/shurcooL/sanitized_anchor_name v1.0.0
```

Instead of what it was previously:

```
require github.com/shurcooL/sanitized_anchor_name v0.0.0-20170918181015-86672fcb3f95
```
